Add initial checkpointing to the nullable walker.

333fred opened the pull request:
This adds the initial code for checkpointing the nullable walker, and defines the structure of a checkpoint. This PR does not do anything with the checkpoints yet: MemberSemanticModel gets the checkpoints, and then just throws them away without saving them. The intention is that it will save the checkpoints and use them in the speculative model in a follow up PR. @dotnet/roslyn-compiler @AlekseyTs @gafter for review.

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4682,6 +4678,17 @@ static TypeWithState calculateResultType(TypeWithAnnotations targetTypeWithNulla
             return operandType;
         }
 
+        private void CheckpointWalkerThroughConversionGroup(BoundConversion conversionOpt, BoundExpression convertedNode)
```
I chose to add this as a separate method instead of putting it in `RemoveConversion` directly, as `RemoveConversion` is often used without then visiting the underlying expression. There are tradeoffs to both though, so there's preference on adding it to `RemoveConversion`, controlled by a default parameter, I can do that as well.

333fred commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.SpeculativeMemberSemanticModel.cs:
```diff
@@ -22,14 +23,14 @@ public SpeculativeMemberSemanticModel(SyntaxTreeSemanticModel parentSemanticMode
                 // Currently, it will return incorrect info because it's just running analysis on the speculated
                 // part.
-                return NullableWalker.AnalyzeAndRewrite(Compilation, MemberSymbol as MethodSymbol, boundRoot, binder, diagnostics);
+                return NullableWalker.AnalyzeAndRewrite(Compilation, MemberSymbol as MethodSymbol, boundRoot, binder, diagnostics, checkpointMapOpt: null);
```
In a follow-up PR I will be asserting that a checkpoint map isn't passed to this version of the semantic model.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs:
```diff
@@ -0,0 +1,164 @@
+                    var pooledDict = PooledDictionary<K, V>.GetInstance();
+                    foreach (var (key, value) in dictOpt)
+                    {
+                        pooledDict[key] = value;
```
>pooledDict[key] = value [](start = 24, length = 23)

Please use `.Add(key, value)` to ensure there are no duplicates.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs:
```diff
@@ -0,0 +1,164 @@
+                ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt = null;
+                if (!_returnTypesOpt.IsDefault)
+                {
+                    returnTypesOpt = ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)>.GetInstance(_returnTypesOpt.Length);
```
>ArrayBuilder [](start = 37, length = 12)

`new ArrayBuilder<...>()` since it doesn't look like this will be freed.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -408,15 +379,17 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             Binder binder,
-            DiagnosticBag diagnostics)
+            DiagnosticBag diagnostics,
+            Dictionary<BoundNode, Checkpoint> checkpointMapOpt)
```
>Dictionary<BoundNode, Checkpoint> checkpointMapOpt [](start = 12, length = 50)

Could this be returned as an `ImmutableDictionary<,>`?

333fred replied:
I expect this method to be used both by the speculative semantic path and non-speculative path, and the speculative path won't need to actually do the checkpointing. I could change this to be a bool flag about whether to do the checkpointing and return null in that case if you'd prefer.

---
In reply to: [286218983](https://github.com/dotnet/roslyn/pull/35850#discussion_r286218983) [](ancestors = 286218983)

cston replied:
Changing to `bool createMap, out ImmutableDictionary<,> map` sounds reasonable.

---
In reply to: [286219930](https://github.com/dotnet/roslyn/pull/35850#discussion_r286219930) [](ancestors = 286219930,286218983)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs:
```diff
@@ -0,0 +1,164 @@
+                    walker._useMethodSignatureParameterTypes,
+                    walker._symbol,
+                    walker._returnTypesOpt?.ToImmutable() ?? default,
+                    walker._methodGroupReceiverMapOpt?.ToImmutableDictionary(),
```
>walker._methodGroupReceiverMapOpt?.ToImmutableDictionary(), [](start = 20, length = 59)

Can we share this `ImmutableDictionary<,>` with all checkpoints, until the original map changes? Same question for other immutable snapshots here.

cston replied:
If we're creating checkpoints at each node, then perhaps the actual `NullableWalker` fields should be immutable objects so the checkpoints can simply copy the fields by reference.

---
In reply to: [286220737](https://github.com/dotnet/roslyn/pull/35850#discussion_r286220737) [](ancestors = 286220737)

333fred replied:
This will force us to create allocations for these new objects in the general straight-line case, which is a tradeoff we can make but I want to be sure we're deliberate about choosing to do so.

---
In reply to: [286227877](https://github.com/dotnet/roslyn/pull/35850#discussion_r286227877) [](ancestors = 286227877,286220737)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs:
```diff
@@ -0,0 +1,164 @@
+        /// Contains a checkpoint for the NullableWalker at any given point of execution, used for restoring the walker to
+        /// a specific point for speculatively analyzing a piece of code that does not appear in the original tree.
+        /// </summary>
+        internal readonly struct Checkpoint
```
>Checkpoint [](start = 33, length = 10)

Should this be named "Snapshot" perhaps? "Checkpoint" sounds like a point in time where a number of checks are performed rather than capturing state.

333fred replied:
I don't know if it's the gamer in me, but in video games a "checkpoint" is the last state you go back to if you fail the level. I can use a different word if you really want, but checkpoint is correct for this as well.

---
In reply to: [286230409](https://github.com/dotnet/roslyn/pull/35850#discussion_r286230409) [](ancestors = 286230409)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+        /// Contains a checkpoint for the NullableWalker at any given point of execution, used for restoring the walker to
+        /// a specific point for speculatively analyzing a piece of code that does not appear in the original tree.
+        /// </summary>
+        internal readonly struct Snapshot
```
>Snapshot [](start = 33, length = 8)

It feels like this structure can be significantly trimmed as we discussed offline.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                    binder.Conversions,
+                    this._variableState,
+                    returnTypesOpt: default,
+                    analyzedNullabilityMap,
```
>analyzedNullabilityMap [](start = 20, length = 22)

Is the walker going to add to the map or to look for something in it?

333fred replied:
It will add to the map. This is how nullability info on the speculative code will be conveyed to the speculative model.

---
In reply to: [288720277](https://github.com/dotnet/roslyn/pull/35850#discussion_r288720277) [](ancestors = 288720277)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                    {
+                        // The first slot always does not exist for the purposes of caching and can be ignored
+                        // in the exit condition check
+                        if (i != 0 && !walkerVariableBySlot[i].Exists)
```
>i != 0 [](start = 28, length = 6)

Consider moving this logic out of the loop and starting at ```i = 1```.

333fred replied:
I'm ripping out this caching logic entirely, so I will close these comments as won't fix as they're no longer relevant.

---
In reply to: [288751142](https://github.com/dotnet/roslyn/pull/35850#discussion_r288751142) [](ancestors = 288751142)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                        // in the exit condition check
+                        if (i != 0 && !walkerVariableBySlot[i].Exists)
+                        {
+                            break;
```
>break; [](start = 28, length = 6)

Can we use ```nextVariableSlot``` instead to figure out required capacity of the builder and to know how many items to copy with AddRange?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                    int i;
+                    for (i = 0; i < _lastVariableBySlot.Length; i++)
+                    {
+                        Debug.Assert(compareIdentifiers(_lastVariableBySlot[i], _walkerToSnapshot.variableBySlot[i]));
```
>Debug.Assert(compareIdentifiers(_lastVariableBySlot[i], _walkerToSnapshot.variableBySlot[i])); [](start = 24, length = 94)

It looks like this code can be simplified if we assert that ```_lastVariableBySlot.Length == nextVariableSlot```

AlekseyTs replied:
Assuming that we handle always empty 0's element outside of the loop.

---
In reply to: [288757319](https://github.com/dotnet/roslyn/pull/35850#discussion_r288757319) [](ancestors = 288757319)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                    Debug.Assert(_lastVariableSlot.Count == _walkerToSnapshot._variableSlot.Count);
+                    foreach ((VariableIdentifier key, int value) in _lastVariableSlot)
+                    {
+                        if (!key.Exists) continue;
```
>continue; [](start = 41, length = 9)

Is this code reachable?

333fred replied:
Yes, slot 0 never exists. However, this code is now gone.

---
In reply to: [288758660](https://github.com/dotnet/roslyn/pull/35850#discussion_r288758660) [](ancestors = 288758660)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+#endif
+            }
+
+            internal void SnapshotWalker(BoundNode node)
```
>SnapshotWalker [](start = 26, length = 14)

Is this function supposed to take a snapshot? The name is very confusing then. Consider renaming.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                _lastConditionalAccessSlot = lastConditionalAccessSlot;
+            }
+
+            internal static Snapshot SnapshotWalker(NullableWalker walker) =>
```
>SnapshotWalker [](start = 37, length = 14)

The name feels very confusing.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+
+            internal static Snapshot SnapshotWalker(NullableWalker walker) =>
+                new Snapshot(
+                    walker._snapshotter.GetVariableState(walker.State),
```
>walker.State [](start = 57, length = 12)

Wouldn't we want to clone the state?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+        /// taking a snapshot of the NullableWalker in order to be as efficient
+        /// as possible when creating many snapshots.
+        /// </summary>
+        private sealed class Snapshotter
```
>Snapshotter [](start = 29, length = 11)

It feels like we should test this machinery one way or another.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -133,7 +104,7 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
         /// Return statements and the result types from analyzing the returned expressions. Used when inferring lambda return type in MethodTypeInferrer.
         /// </summary>
-        private readonly ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> _returnTypesOpt;
+        private ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> _returnTypesOpt;
```
>private ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> _returnTypesOpt; [](start = 8, length = 84)

I do not believe there is any advantage to change the type of this field to immutable array.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -147,14 +118,19 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
         /// (Could be replaced by _analyzedNullabilityMapOpt if that map is always available.)
         /// </summary>
-        private PooledDictionary<BoundExpression, TypeWithState> _methodGroupReceiverMapOpt;
+        private ImmutableDictionary<BoundExpression, TypeWithState> _methodGroupReceiverMapOpt;
```
>private ImmutableDictionary<BoundExpression, TypeWithState> _methodGroupReceiverMapOpt; [](start = 8, length = 87)

I do not believe there is any advantage in changing the type of this field.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -285,7 +261,7 @@ private void SetAnalyzedNullability(BoundExpression expr, VisitResult result, bo
         /// Placeholder locals, e.g. for objects being constructed.
         /// </summary>
-        private PooledDictionary<object, PlaceholderLocal> _placeholderLocalsOpt;
+        private ImmutableDictionary<object, PlaceholderLocal> _placeholderLocalsOpt;
```
>private ImmutableDictionary<object, PlaceholderLocal> _placeholderLocalsOpt; [](start = 8, length = 76)

I do not believe there is any advantage in changing the type of this field.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -315,9 +283,11 @@ protected override void Free()
-            ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt,
             VariableState initialState,
-            Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt)
+            ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt,
```
>ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt [](start = 12, length = 74)

What is the point in passing this array in? Would we ever use its content during analysis?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -339,12 +309,24 @@ protected override void Free()
-                foreach (var pair in initialState.VariableTypes)
+
+                var variableTypesBuilder = PooledDictionary<Symbol, TypeWithAnnotations>.GetInstance();
+                foreach (var (key, value) in initialState.VariableTypes)
```
>initialState.VariableTypes [](start = 45, length = 26)

Isn't this already an immutable dictionary? Can't we just use it as is?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                }
+            }
+
+            internal ImmutableDictionary<BoundNode, Snapshot> Snapshots => _snapshotMap.ToImmutableDictionary();
```
>_snapshotMap [](start = 75, length = 12)

Should we handle ```null``` here?

333fred replied:
No longer necessary: if there's a SnapshotManager, there is no null chance.

---
In reply to: [288779165](https://github.com/dotnet/roslyn/pull/35850#discussion_r288779165) [](ancestors = 288779165)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -315,9 +283,11 @@ protected override void Free()
+            ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt,
+            Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt,
+            bool takeIncrementalSnapshots,
+            Snapshotter snapshotter = null)
```
I think it would be better to not add both parameters to this method, otherwise how would we make sure they agree with each other.

333fred replied:
Initially I had no choice, as the Snapshotter was used both when taking incremental snapshots and for getting the variable state to pass to the child method. That has been factored out and we just pass the manager.

---
In reply to: [288780361](https://github.com/dotnet/roslyn/pull/35850#discussion_r288780361) [](ancestors = 288780361)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -369,18 +351,27 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
             ParameterSymbol methodThisParameter = MethodThisParameter;
             this.regionPlace = RegionPlace.Before;
-            EnterParameters(); // assign parameters
+            var parametersBuilder = PooledDictionary<Symbol, TypeWithAnnotations>.GetInstance();
```
>PooledDictionary<Symbol, TypeWithAnnotations>.GetInstance(); [](start = 36, length = 60)

Would it make sense to use a builder specialized for ImmutableDictionary?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -369,18 +351,27 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
-                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations);
+                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations, parametersBuilder);
             }
+            _variableTypes = parametersBuilder.ToImmutableDictionaryAndFree();
```
>_variableTypes = parametersBuilder.ToImmutableDictionaryAndFree(); [](start = 12, length = 66)

Is the motivation for using the builder to add/update several entries in bulk? Is it worth the complexity?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -369,18 +351,27 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
+            _variableTypes = parametersBuilder.ToImmutableDictionaryAndFree();
+
+            // We need to do a checkpoint even of the first node, because we want to have the state of the initial parameters.
+            _snapshotter.SnapshotWalker(methodMainNode);
```
>_snapshotter.SnapshotWalker(methodMainNode); [](start = 12, length = 44)

When we are taking a snapshot for a bound node, it it a snapshot of a state before we visit the node or after the node has been visited? In any case, I do not feel comfortable about this special case for the root bound node.

333fred replied:
It's the state before the node. With us moving to position based, are you still uncomfortable with this? We need to take a snapshot of the initial state of the parameters at the start of the method body .

---
In reply to: [288784061](https://github.com/dotnet/roslyn/pull/35850#discussion_r288784061) [](ancestors = 288784061)

AlekseyTs replied:
> With us moving to position based, are you still uncomfortable with this?

I think position-wise that would make sense.

---
In reply to: [288815668](https://github.com/dotnet/roslyn/pull/35850#discussion_r288815668) [](ancestors = 288815668,288784061)

333fred replied:
Right, and the first position of the method is the main body node, at which point all of the state we just set above (namely the initial states of the parameters) will need to be populated in a snapshot.

---
In reply to: [289073267](https://github.com/dotnet/roslyn/pull/35850#discussion_r289073267) [](ancestors = 289073267,288815668,288784061)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -429,19 +428,27 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             }
 #endif
+
+            snapshotManager.Free();
```
>snapshotManager [](start = 12, length = 15)

Is this never null?

333fred replied:
No longer relevant.

---
In reply to: [288785768](https://github.com/dotnet/roslyn/pull/35850#discussion_r288785768) [](ancestors = 288785768)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -465,20 +474,48 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
-                analyzedNullabilityMapOpt: null);
+                analyzedNullabilityMapOpt: null,
+                takeIncrementalSnapshots: false,
+                snapshotManagerOpt: ref ignored,
```
>ignored [](start = 40, length = 7)

Should we ```Free``` this?

333fred replied:
No longer relevant.

---
In reply to: [288785918](https://github.com/dotnet/roslyn/pull/35850#discussion_r288785918) [](ancestors = 288785918)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -503,11 +542,14 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
-            Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt)
+            Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt,
+            bool takeIncrementalSnapshots,
+            ref Snapshotter snapshotManagerOpt,
```
>ref Snapshotter snapshotManagerOpt [](start = 12, length = 34)

Passing this by ref feels too complicated.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -754,6 +801,31 @@ static MethodSymbol getTopLevelMethod(MethodSymbol method)
+            {
+                builder.Add(newKey, newValue);
+            }
+            toUpdate = builder.ToImmutableDictionaryAndFree();
```
>toUpdate = builder.ToImmutableDictionaryAndFree(); [](start = 12, length = 50)

This looks unnecessary complex. Could we use ```Set​Item```/```Add``` methods instead?

333fred replied:
I missed the `SetItem` method.

---
In reply to: [288787926](https://github.com/dotnet/roslyn/pull/35850#discussion_r288787926) [](ancestors = 288787926)

333fred replied:
Turns out we can't use `SetItem` with TypeSymbols, that ends up calling into its implementation of `Equals(object)`, which throws. I've moved `_variableTypes` back to a `PooledDictionary`

---
In reply to: [288815798](https://github.com/dotnet/roslyn/pull/35850#discussion_r288815798) [](ancestors = 288815798,288787926)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+                _walkerToSnapshot = walker;
+                if (takeIncrementalSnapshot)
+                {
+                    _snapshotMap = PooledDictionary<BoundNode, Snapshot>.GetInstance();
```
>_snapshotMap = PooledDictionary<BoundNode, Snapshot>.GetInstance(); [](start = 20, length = 67)

It feels like it would be better to have this map be always supplied by a consumer. Consumer owns the map and is responsible for freeing it and turn to immutable when appropriate.

AlekseyTs replied:
This way sharing the same map across different Snapshotter instances will be more natural and straightforward.

---
In reply to: [288789402](https://github.com/dotnet/roslyn/pull/35850#discussion_r288789402) [](ancestors = 288789402)

333fred replied:
I'm moving this to a more standard builder model, and I don't think this comment is relevant anymore.

---
In reply to: [288798291](https://github.com/dotnet/roslyn/pull/35850#discussion_r288798291) [](ancestors = 288798291,288789402)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+            private readonly NullableWalker _walkerToSnapshot;
+            private ImmutableDictionary<VariableIdentifier, int> _lastVariableSlot = ImmutableDictionary<VariableIdentifier, int>.Empty;
+            private ImmutableArray<VariableIdentifier> _lastVariableBySlot = ImmutableArray<VariableIdentifier>.Empty;
+            private readonly PooledDictionary<BoundNode, Snapshot> _snapshotMap;
```
>PooledDictionary<BoundNode, Snapshot> [](start = 29, length = 37)

Would it be better to use a builder specialized for ImmutableDictionary?

AlekseyTs replied:
>BoundNode [](start = 46, length = 9)

Since this is meant to serve speculative SemanticModel APIs and those use positions to identify the place to speculate at, it feels like it would be more natural to use position as a key in this map as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs:
```diff
@@ -0,0 +1,250 @@
+            {
+                if (_snapshotMap != null)
+                {
+                    _snapshotMap[node] = Snapshot.SnapshotWalker(_walkerToSnapshot);
```
> _snapshotMap[node] = Snapshot.SnapshotWalker(_walkerToSnapshot); [](start = 19, length = 65)

Is there a way to assert that we are not overwriting values during the same pass?

AlekseyTs replied:
>node [](start = 33, length = 4)

We probably shouldn't take any snapshots for compiler generated bound nodes.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs:
```diff
@@ -57,6 +56,12 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
 
             public override BoundNode Visit(BoundNode node)
             {
+                // Ensure that we always have a checkpoint for every BoundExpression in the map
```
>BoundExpression [](start = 69, length = 15)

Is the comment accurate? I do not see anything specific for BoundExpression nodes.

cston replied:
>checkpoint [](start = 48, length = 10)

"snapshot"

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
@AlekseyTs feedback has been addressed.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+            internal NullableWalker RestoreWalkerToAnalyzeNewNode(
+                int position,
+                BoundNode nodeToAnalyze,
+                CSharpCompilation compilation,
```
>CSharpCompilation compilation, [](start = 16, length = 30)

Do we need to pass this in? Is the binder always available? I think we can get compilation from the binder.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                bool foundSnapshot = false;
+                foreach (var (currentPosition, currentSnapshot) in _incrementalSnapshots)
+                {
+                    if (currentPosition <= position)
```
>if (currentPosition <= position) [](start = 20, length = 32)

Are we iterating sequence in any particular order? If not, then it looks like we likely to grab a snapshot in a non-deterministic way.

333fred replied:
Will switch to `ImmutableSortedDictionary` for now. We do need to revamp this search code entirely, but that will at least get us in a consistent state.

---
In reply to: [289474582](https://github.com/dotnet/roslyn/pull/35850#discussion_r289474582) [](ancestors = 289474582)

AlekseyTs replied:
It looks like this implementation is always going to pick a snapshot at the beginning of the method body.

---
In reply to: [289501930](https://github.com/dotnet/roslyn/pull/35850#discussion_r289501930) [](ancestors = 289501930,289474582)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+            {
+                private readonly ArrayBuilder<SharedWalkerState> _walkerStates = ArrayBuilder<SharedWalkerState>.GetInstance();
+                private readonly ImmutableDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableDictionary.CreateBuilder<int, Snapshot>();
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
```
>private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance(); [](start = 16, length = 107)

Consider adding a doc comment. What symbols do we map to what slots, etc.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                private readonly ArrayBuilder<SharedWalkerState> _walkerStates = ArrayBuilder<SharedWalkerState>.GetInstance();
+                private readonly ImmutableDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableDictionary.CreateBuilder<int, Snapshot>();
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private readonly ArrayBuilder<int> _previousWalkerSlots = ArrayBuilder<int>.GetInstance();
```
>private readonly ArrayBuilder<int> _previousWalkerSlots = ArrayBuilder<int>.GetInstance(); [](start = 16, length = 90)

Consider adding a doc comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1 && _previousWalkerSlots.Count == 0, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_walkerStates.Count == _nextWalkerSlot);
```
>Debug.Assert(_walkerStates.Count == _nextWalkerSlot); [](start = 20, length = 53)

Is there an advantage to have a separate field if we can always use ```_walkerStates.Count``` to get the value?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1 && _previousWalkerSlots.Count == 0, "Attempting to finalize snapshots before all walks completed");
```
>_previousWalkerSlots.Count == 0 [](start = 61, length = 31)

It looks like we can split this assert into two independent asserts.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+
+                internal void EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(!(symbol is null));
```
>!(symbol is null) [](start = 33, length = 17)

```symbol is object```?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                    // Because we potentially run multiple passes, we
+                    // need to make sure we use the same final shared
+                    // state for following passes.
+                    if (_symbolToSlot.ContainsKey(symbol))
```
>_symbolToSlot.ContainsKey(symbol) [](start = 24, length = 33)

Please use TryGetValue to avoid doing lookup twice

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                    }
+                    else
+                    {
+                        _currentWalkerSlot = _nextWalkerSlot;
```
>_currentWalkerSlot = _nextWalkerSlot; [](start = 24, length = 37)

This feels fragile, consider using ```_symbolToSlot.Count``` instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+                internal void EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(!(symbol is null));
+                    _previousWalkerSlots.Push(_currentWalkerSlot);
```
>_previousWalkerSlots.Push(_currentWalkerSlot); [](start = 20, length = 46)

We can avoid allocating a special stack for that by simply returning previous slot to the caller and passing it back into ExitWalker.

333fred replied:
We could, but I would really prefer to keep the knowledge of the walker slots encapuslated within the builder and manager. I'll change it if you insist, but that's a code smell to me.

---
In reply to: [289483513](https://github.com/dotnet/roslyn/pull/35850#discussion_r289483513) [](ancestors = 289483513)

cston replied:
I agree with @AlekseyTs - the previous symbol should be returned to the caller.

---
In reply to: [289492447](https://github.com/dotnet/roslyn/pull/35850#discussion_r289492447) [](ancestors = 289492447,289483513)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -465,20 +486,46 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
-            ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> returnTypes,
             VariableState initialState,
-            Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt)
+            ImmutableDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)>.Builder analyzedNullabilityMapOpt)
```
>ImmutableDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)>.Builder analyzedNullabilityMapOpt [](start = 12, length = 101)

Is there any consumer that uses this parameter? Consider removing otherwise.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -754,6 +819,9 @@ static MethodSymbol getTopLevelMethod(MethodSymbol method)
             }
         }
 
+        private void AddDeclaredVariable(Symbol symbol, TypeWithAnnotations declaredType) =>
```
>AddDeclaredVariable [](start = 21, length = 19)

Consider removing.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4682,6 +4766,19 @@ static TypeWithState calculateResultType(TypeWithAnnotations targetTypeWithNulla
             return operandType;
         }
 
+        private void SnapshotWalkerThroughConversionGroup(BoundConversion conversionOpt, BoundExpression convertedNode)
```
>SnapshotWalkerThroughConversionGroup [](start = 21, length = 36)

Is this gymnastics still needed after the switch to position based snapshots?

333fred replied:
Yes, because there are a couple of cases where we remove both explicit and implicit conversions. I now bail out as soon as we get to a conversion with the same syntax as the underlying node to avoid taking extra snapshots.

---
In reply to: [289493249](https://github.com/dotnet/roslyn/pull/35850#discussion_r289493249) [](ancestors = 289493249)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4944,9 +5042,10 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
                         initialState: GetVariableState(this.TopState()),
-                        analyzedNullabilityMapOpt: _disableNullabilityAnalysis ? null : _analyzedNullabilityMapOpt);
+                        analyzedNullabilityMapOpt: _disableNullabilityAnalysis ? null : _analyzedNullabilityMapOpt,
+                        _snapshotBuilderOpt,
```
>_snapshotBuilderOpt [](start = 24, length = 19)

Should this be conditional based on ```_disableNullabilityAnalysis```?

cston commented on src/Compilers/CSharp/Portable/Compilation/AttributeSemanticModel.cs:
```diff
@@ -98,9 +98,9 @@ internal override BoundNode Bind(Binder binder, CSharpSyntaxNode node, Diagnosti
         }
 
-        protected override BoundNode RewriteNullableBoundNodes(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics)
+        protected override BoundNode RewriteNullableBoundNodesWithCheckpoints(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, out NullableWalker.SnapshotManager snapshotManager)
```
>Checkpoints [](start = 66, length = 11)

`Snapshots`?

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1850,15 +1850,15 @@ private void EnsureRootBoundForNullabilityIfNecessary()
         }
 
-        protected abstract BoundNode RewriteNullableBoundNodes(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics);
+        protected abstract BoundNode RewriteNullableBoundNodesWithCheckpoints(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, out NullableWalker.SnapshotManager snapshotManager);
```
>Checkpoints [](start = 66, length = 11)

`Snapshots`?

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1850,15 +1850,15 @@ private void EnsureRootBoundForNullabilityIfNecessary()
             var binder = GetEnclosingBinder(GetAdjustedNodePosition(bindableRoot));
             var boundRoot = Bind(binder, bindableRoot, diagnostics);
-            boundRoot = RewriteNullableBoundNodes(boundRoot, binder, diagnostics);
+            boundRoot = RewriteNullableBoundNodesWithCheckpoints(boundRoot, binder, diagnostics, out var checkpointMap);
```
>var checkpointMap [](start = 101, length = 17)

`_` or `var snapshotManager`

333fred replied:
I'll use `_` for now. We'll consume it later.

---
In reply to: [289510893](https://github.com/dotnet/roslyn/pull/35850#discussion_r289510893) [](ancestors = 289510893)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -382,6 +390,9 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
                 EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations);
             }
 
+            // We need to do a checkpoint even of the first node, because we want to have the state of the initial parameters.
```
>to do a checkpoint [](start = 23, length = 18)

"to create a snapshot"

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -406,20 +417,27 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+            var analyzedNullabilities = ImmutableDictionary.CreateBuilder<BoundExpression, (NullabilityInfo, TypeSymbol)>();
             var methodSymbol = symbol as MethodSymbol;
+            // Attributes don't have a symbol, which is what SnapshotBuilder uses as an index for maintaining global state.
+            // Until we have a workaround for this, disable snapshotting for null symbols.
```
>snapshotting [](start = 60, length = 12)

"snapshots"

AlekseyTs commented:
Done with review pass (iteration 5)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -465,20 +486,46 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         }
 
-        internal static void Analyze(
+        internal static ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> Analyze(
```
>Analyze [](start = 84, length = 7)

Is this overload necessary, or could callers call into the other overload that takes an `ArrayBuilder` and avoid creating `ImmutableArray` instances?

333fred replied:
The only consumers of this overload use the produced `ImmutableArray`. I don't think it's a cost.

---
In reply to: [289515254](https://github.com/dotnet/roslyn/pull/35850#discussion_r289515254) [](ancestors = 289515254)

cston replied:
The consumers previously used `ArrayBuilder`. Is there a reason to change that?

---
In reply to: [289538121](https://github.com/dotnet/roslyn/pull/35850#discussion_r289538121) [](ancestors = 289538121,289515254)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,183 @@
+        /// <summary>
+        /// Contains the shared state state used to restore the walker at a specific point
+        /// </summary>
+        internal struct SharedWalkerState
```
>SharedWalkerState [](start = 24, length = 17)

Is this different from the `VariableState` type?

333fred replied:
Yes. `VariableState` does not hold onto the `Symbol`, and it does hold onto a `LocalState`.

---
In reply to: [289521654](https://github.com/dotnet/roslyn/pull/35850#discussion_r289521654) [](ancestors = 289521654)

333fred commented:
@AlekseyTs @cston addressed your feedback.

---
In reply to: [497824008](https://github.com/dotnet/roslyn/pull/35850#issuecomment-497824008) [](ancestors = 497824008)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_symbolToSlot.Count == _walkerStates.Count);
```
>_symbolToSlot.Count [](start = 33, length = 19)

Would it make sense to assert that ```_symbolToSlot.Count > 0```?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+#if DEBUG
+                    Debug.Assert(symbol is object);
+                    Debug.Assert((_currentSymbol is null && _currentWalkerSlot == -1 && _initialSymbol is null) ||
+                                 symbol.ContainingSymbol.Equals(_currentSymbol));
```
>symbol.ContainingSymbol.Equals(_currentSymbol) [](start = 33, length = 46)

Should be able to use reference equality here

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+                internal void ExitWalker(SharedWalkerState stableState, Symbol symbol)
+                {
+#if DEBUG
+                    Debug.Assert(_currentSymbol.Equals(symbol));
```
>_currentSymbol.Equals(symbol) [](start = 33, length = 29)

Should be able to use reference equality here

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+                    _currentSymbol = symbol.ContainingSymbol;
+#endif
+                    _walkerStates.SetItem(_currentWalkerSlot, stableState);
+                    if (_initialSymbol.Equals(symbol))
```
>_initialSymbol.Equals(symbol) [](start = 24, length = 29)

Should be able to use reference equality here

AlekseyTs replied:
>_initialSymbol.Equals(symbol) [](start = 24, length = 29)

I think we can get rid of ```_initialSymbol``` field. Here we can check if ```_currentWalkerSlot == 0``` instead and it doesn't look like it adds much value elsewhere.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+                    if (_initialSymbol.Equals(symbol))
+                    {
+                        Debug.Assert(_currentWalkerSlot == 0);
+                        _currentWalkerSlot = -1;
```
>_currentWalkerSlot = -1; [](start = 24, length = 24)

It feels like we would want to reset _currentSymbol to null as well.

AlekseyTs replied:
And _initialSymbol   too.

---
In reply to: [289559721](https://github.com/dotnet/roslyn/pull/35850#discussion_r289559721) [](ancestors = 289559721)

AlekseyTs commented:
Done with review pass (iteration 6)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,210 @@
+                    _currentSymbol = symbol;
+#endif
+                    _initialSymbol ??= symbol;
+                    var previousSlot = _currentWalkerSlot;
```
>previousSlot [](start = 24, length = 12)

Not used.

333fred commented:
@AlekseyTs @cston addressed feedback.

---
In reply to: [497871089](https://github.com/dotnet/roslyn/pull/35850#issuecomment-497871089) [](ancestors = 497871089)

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+
+                internal int EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(symbol is object);
```
@AlekseyTs unfortunately the assertion that this incoming `Symbol` is related to the current symbol proved to not be true in bad code scenarios like this one: https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs#L1540-L1555.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -108,12 +108,12 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
                 // will be called again from NullableWalker.ApplyConversion when the
                 // BoundLambda is converted to an anonymous function.
                 // https://github.com/dotnet/roslyn/issues/31752: Can we avoid generating extra
-                // diagnostics? And is this exponential when there are nested lambdas?
```
>? [](start = 85, length = 1)

Dropped a "?"

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4682,6 +4739,19 @@ static TypeWithState calculateResultType(TypeWithAnnotations targetTypeWithNulla
 
+        private void SnapshotWalkerThroughConversionGroup(BoundConversion conversionOpt, BoundExpression convertedNode)
+        {
+            var conversionGroup = conversionOpt?.ConversionGroupOpt;
```
>var [](start = 12, length = 3)

Consider returning early if `_snapshotBuilderOpt is null`.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4682,6 +4739,19 @@ static TypeWithState calculateResultType(TypeWithAnnotations targetTypeWithNulla
             return operandType;
         }
 
+        private void SnapshotWalkerThroughConversionGroup(BoundConversion conversionOpt, BoundExpression convertedNode)
```
>BoundConversion conversionOpt [](start = 58, length = 29)

All callers pass `expr as BoundConversion`. Consider changing this argument to `BoundExpression` and checking for `BoundConversion` here. (We're already performing that check inside the `while`.)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+        }
+
+        /// <summary>
+        /// Contains the shared state state used to restore the walker at a specific point
```
>state state [](start = 32, length = 11)

Typo

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+        /// <summary>
+        /// Contains the shared state state used to restore the walker at a specific point
+        /// </summary>
+        internal struct SharedWalkerState
```
>Shared [](start = 24, length = 6)

Is "Shared" necessary in the name or comment above?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+        private readonly struct Snapshot
+        {
+            internal readonly LocalState VariableState;
+            internal readonly int GlobalStateIndex;
```
>Global [](start = 34, length = 6)

Consider using "Shared" rather than introducing a different term.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:
```diff
@@ -493,7 +503,10 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
             foreach (var arm in node.SwitchArms)
             {
                 SetState(!arm.Pattern.HasErrors && labelStateMap.TryGetValue(arm.Label, out var labelState) ? labelState.state : UnreachableState());
+                // https://github.com/dotnet/roslyn/issues/35836 Is this where we want to take the checkpoint?
```
>checkpoint [](start = 99, length = 10)

"snapshot"

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+            /// <summary>
+            /// The int key corresponds to <see cref="Snapshot.GlobalStateIndex"/>.
+            /// </summary>
+            private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
```
>ImmutableArray<SharedWalkerState> _walkerGlobalStates [](start = 29, length = 53)

Is this array necessary? If `SharedWalkerState` were a `class`, could we use a reference to `SharedWalkerState` in `Snapshot` rather than an index into this array?

333fred replied:
We could potentially do that, but I'm going to leave this as is. I think making `SharedWalkerState` a class would probably introduce more garbage than it would save.

---
In reply to: [289925758](https://github.com/dotnet/roslyn/pull/35850#discussion_r289925758) [](ancestors = 289925758)

cston replied:
There shouldn't be many allocations - roughly one instance of `SharedWalkerState` per variable.

But if that is a concern, the fields of `SharedWalkerState` could be copied into `Snapshot`.

---
In reply to: [289940273](https://github.com/dotnet/roslyn/pull/35850#discussion_r289940273) [](ancestors = 289940273,289925758)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,202 @@
+            /// closest position before a given position.
+            /// </summary>
+            private sealed class DescendingIntComparer : IComparer<int>
+            {
```
>{ [](start = 12, length = 1)

Consider adding a singleton instance.

333fred commented:
@AlekseyTs @cston addressed feedback.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,203 @@
+            /// <summary>
+            /// The int key corresponds to <see cref="Snapshot.SharedStateIndex"/>.
+            /// </summary>
+            private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
```
>Global [](start = 70, length = 6)

"Shared"?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,203 @@
+                    return null;
+                }
+
+                var globalState = _walkerGlobalStates[incrementalSnapshot.SharedStateIndex];
```
>globalState [](start = 20, length = 11)

"sharedState"?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -0,0 +1,203 @@
+            internal readonly LocalState VariableState;
+            internal readonly int SharedStateIndex;
+
+            internal Snapshot(LocalState variableState, int globalStateIndex)
```
>global [](start = 60, length = 6)

"shared"

333fred commented:
@cston I'll address those comments in the next followup PR.
